docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,275 @@
+	// of pods stable.
+	DesiredReplicas int
+
+	// CurrentConsumption is the current average consumption of the given resource that the autoscaler will
+	// try to maintain by adjusting the desired number of pods.
+	// Two types of resources are supported: "cpu" and "memory".
+	CurrentConsumption ResourceConsumption
+
+	// LastScaleTimestamp is the last time the HorizontalPodAutoscaler scaled the number of pods.
+	// This is used by the autoscaler to controll how often the number of pods is changed.
+	LastScaleTimestamp *util.Time
+}
+
+// ResourceConsumption is an object for specifying average resource consumption of a particular resource.
+type ResourceConsumption struct {
+	Resource api.ResourceName
+	Quantity resource.Quantity
+}
+```
+
+```Scale``` will be a reference to the Scale subresource.
+```MinCount```, ```MaxCount``` and ```Target``` will define autoscaler configuration.
+We will also introduce HorizontalPodAutoscalerList object to enable listing all autoscalers in the cluster:
+
+```go
+// HorizontalPodAutoscaler is a collection of pod autoscalers.
+type HorizontalPodAutoscalerList struct {
+	api.TypeMeta
+	api.ListMeta
+
+	Items []HorizontalPodAutoscaler
+}
+```
+
+## Autoscaling Algorithm
+
+The autoscaler will be implemented as a control loop.
+It will periodically (e.g.: every 1 minute) query pods described by ```Status.PodSelector``` of Scale subresource,
+and check their average CPU or memory usage from the last 1 minute
+(there will be API on master for this purpose, see
+[#11951](https://github.com/GoogleCloudPlatform/kubernetes/issues/11951).
+Then, it will compare the current CPU or memory consumption with the Target,
+and adjust the count of the Scale if needed to match the target
+(preserving condition: MinCount <= Count <= MaxCount).
+
+The target number of pods will be calculated from the following formula:
+
+```
+TargetNumOfPods = sum(CurrentPodsConsumption) / Target
+```
+
+To make scaling more stable, scale-up will happen only when the floor of ```TargetNumOfPods``` is higher than
+the current number, while scale-down will happen only when the ceiling of ```TargetNumOfPods``` is lower than
+the current number.
+
+The decision to scale-up will be executed instantly.
+However, we will execute scale-down only if the sufficient time has past from the last scale-up (e.g.: 10 minutes).
+Such approach has two benefits:
+
+* Autoscaler works in a conservative way.
+  If new user load appears, it is important for us to rapidly increase the number of pods,
+  so that user requests will not be rejected.
+  Lowering the number of pods is not that urgent.
+
+* Autoscaler avoids thrashing, i.e.: prevents rapid execution of conflicting decision if the load is not stable.
+
+
+As the CPU consumption of a pod immediately after start may be highly variable due to initialization/startup,
+autoscaler will skip metrics from the first minute of pod lifecycle.
```

Comment:
I think 1 minute is a reasonable value to start with in initial version. Later, if it is needed, we may make it configurable.